### PR TITLE
[REFACTOR] Move TBlob and TShape to mxnet

### PR DESCRIPTION
**Pull request by tqchen**
This PR moves TBlob and TShape directly to mxnet. 

The reason of the change was that TBlob does not usually carries computations and having it in mxnet can make the data structure more clear for the users, and helps us on future re-factoring to isolate the components.

Tested this change on ubuntu

### CODE REVIEW

**Comment by tqchen** [comment]
@sxjscience @antinucleon @piiswrong

**Comment by tqchen** [comment]
git push

File: .gitignore [hunk 1]
```diff
@@ -115,3 +115,4 @@ scala-package/*/*/target/
 .project
 .cproject
 .pydevproject
+nnvm
```
Comment (piiswrong) [inline .gitignore]: shouldn't this be a submodule?

File: include/mxnet/tensor_blob.h [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+#define MXNET_TENSOR_BLOB_H_
+
+#include <vector>
+#include <iostream>
+#include <utility>
+#include <algorithm>
+#include <dmlc/logging.h>
+#include "./base.h"
+
+namespace mxnet {
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: The namespace change is going to cause a lot of trouble

File: include/mxnet/tensor_blob.h [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+  Tuple() = default;
+  /*! \brief destructor */
+  inline ~Tuple() {
+    delete [] data_heap_;
+  }
+  /*!
+   * \brief copy constructor from another tuple
+   * \param s the source tuple
+   */
+  inline Tuple(const Tuple<ValueType>& s) {
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: Modify parameter parsing to use tuple instead of tshape?

**Comment by tqchen** [comment]
moved to https://github.com/dmlc/mxnet/pull/2906

File: include/mxnet/tensor_blob.h (hunk 3 above)
Comment (tqchen) [inline include/mxnet/tensor_blob.h]: Template matching for tuple was harder, so stick with TShape for now

File: include/mxnet/tensor_blob.h (hunk 2 above)
Comment (tqchen) [inline include/mxnet/tensor_blob.h]: This is likely to be fine, as originally we typedefed TShape inside mxnet namespace

File: .gitignore (hunk 1 above)
Comment (tqchen) [inline .gitignore]: this is not yet merged in.

**Pull request closed without merging**